lucille-core/src/main/java/com/kmwllc/lucille/stage/DictionaryLookup.java

```diff
@@ -70,9 +88,13 @@ public void start() throws StageException {
    * @return the populated HashMap
    */
   private HashMap<String, String[]> buildHashMap(String dictPath) throws StageException {
-    HashMap<String, String[]> dict = new HashMap<>();
-    try (CSVReader reader = new CSVReader(FileUtils.getReader(dictPath))) {
-      // For each line of the dictionary file, add a keyword/payload pair to the Trie
+
+    // count lines and create a set with correct capacity
+    int lineCount = countLines(dictPath);
```

Comment:
Please add a comment about why & when it's important to set the initial capacity. Does it have a significant performance impact when creating a very large hashset? How large was the hashset that motivated this?

Reply:
thanks for adding the comment -- but I'm curious -- did you observe a specific performance improvement when setting the initial capacity, or did you add the line-counting code just assuming it would be a good idea? Fine either way, I'm just looking for a bit more info.

Reply:
Code:
```java
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.*;

public class Main {
  public static void main(String[] args) {

    List<Long> timesNoCapacity = new ArrayList<>();
    List<Long> timesCapacity = new ArrayList<>();

    for (int i = 1; i < 9; i ++) {

      int size = (int) Math.pow(10, i);

      String filename = "test" + size + ".txt";
      writeRandomInts(size, filename);

      // no capacity
      long start1 = System.currentTimeMillis();
      HashSet<String> set1 = new HashSet<>();
      addValuesFromFile(filename, set1);
      timesNoCapacity.add(System.currentTimeMillis() - start1);

      // clear
      set1.clear();

      // capacity
      long start2 = System.currentTimeMillis();
      HashSet<String> set2 = new HashSet<>((int) (size / .75f) + 1);
      addValuesFromFile(filename, set2);
      timesCapacity.add(System.currentTimeMillis() - start2);

      // clear
      set2.clear();
    }


    System.out.println("size,timesNoCapacity,timesCapacity");
    for (int i = 0; i < timesCapacity.size(); i ++) {
      System.out.println("size=" + ((int) Math.pow(10, i + 1)) + " -> " +  timesNoCapacity.get(i) + ", " + timesCapacity.get(i));
    }

    System.out.println("done");
  }


  // writes n random integers to file
  public static void writeRandomInts(int n, String filename) {
    try {
      FileWriter fw = new FileWriter(filename);
      BufferedWriter bw = new BufferedWriter(fw);
      for (int i = 0; i < n; i ++) {
        bw.write(UUID.randomUUID().toString());
        bw.newLine();
      }
      bw.close();
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  // add values from file to a given set
  public static void addValuesFromFile(String filename, Set<String> set) {
    try {
      Scanner scanner = new Scanner(new File(filename));
      while (scanner.hasNextLine()) {
        set.add(scanner.nextLine());
      }
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }
}
```


Output: 
```
size,timesNoCapacity,timesCapacity
size=10 -> 11, 1
size=100 -> 1, 1
size=1000 -> 6, 5
size=10000 -> 42, 19
size=100000 -> 77, 41
size=1000000 -> 500, 419
size=10000000 -> 5266, 4311
size=100000000 -> 109685, 83468
```

Reply:
This is an awesome comparison -- thank you! -- but, what does "timesCapacity" look like if you include the time to read all the file lines in an initial pass to find the number of lines, rather than knowing the number of lines in advance?

Reply:
updated the code in the comment above.
results with reading lines:
```
size,timesNoCapacity,timesCapacity
size=10 -> 15, 1
size=100 -> 1, 2
size=1000 -> 4, 5
size=10000 -> 13, 8
size=100000 -> 62, 52
size=1000000 -> 541, 487
size=10000000 -> 5268, 4823
size=100000000 -> 106658, 94633
```
Looks like a consistent 10% processing time decrease even with reading the number of lines first

Reply:
That's a cool finding -- thanks for testing this. Could you update the code comment to mention this conclusion -- say that we've observed a 10% time reduction for various sizes above 1000 when setting the initial capacity based on the number of file lines.